### PR TITLE
Fix rust compilation of `openssl-sys`

### PR DESCRIPTION
**Pull request by coffee-cup**
## What does this PR address?

Fix building of Rust projects that use the `openssl-sys` trait. A few things were needed
- Do not build for the MUSL target if openssl is detected
- Set the `OPENSSL_DIR` and `OPENSSL_LIB_DIR` environment variables to the correct locations

The solution mainly comes from a few resources
- https://github.com/PierreZ/nix-env-dev/blob/9dcb33911361b3184af072797baffd340beb0ef7/rust.nix
- https://discourse.nixos.org/t/rust-openssl-woes/12340/3
- https://discourse.nixos.org/t/openssl-dependency-for-rust/3186/3

The `environment.nix` file generation has custom implementation to handle this when the `openssl` Nix library is detected. Ideally this is made generic in the future where providers can set variables based on Nix package locations. But I will think about what that API looks like and possibly add in https://github.com/railwayapp/nixpacks/pull/300. For now this works.

To parse the `Cargo.toml` file in a more accurate way, I am using the `cargo_toml` crate.

Fixes https://github.com/railwayapp/nixpacks/issues/286

### CODE REVIEW

File: src/nixpacks/nix/mod.rs [hunk 1]
```diff
@@ -37,6 +37,17 @@ pub fn create_nix_expression(plan: &BuildPlan) -> String {
...
 
+    // If the openssl library is added, set the OPENSSL_DIR and OPENSSL_LIB_DIR environment variables
+    // In the future, we will probably want a generic way for providers to set variables based off Nix package locations
+    let openssl_dirs = if libraries.contains("openssl") {
+        formatdoc! {"
+      export OPENSSL_DIR=\"${{openssl.dev}}\"
+      export OPENSSL_LIB_DIR=\"${{openssl.out}}/lib\"
+    "}
+    } else {
+        "".to_string()
```
Comment (Milo123459) [inline src/nixpacks/nix/mod.rs:48]: nit: string::new()

**Review by Milo123459 — CHANGES_REQUESTED** [review]
Other than these things, 🙏

File: examples/rust-openssl/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "rust-open-ssl"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+openssl = "0.10.41"
+openssl-sys = "0.9.75"
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply examples/rust-openssl/Cargo.toml:10]: Why not? It is good to test that both deps compile

File: examples/rust-openssl/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+fn main() {
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply examples/rust-openssl/src/main.rs:1]: Good idea

File: examples/rust-openssl/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Milo123459) [reply examples/rust-openssl/Cargo.toml:10]: Openssl depends on openssl-sys

File: src/providers/rust.rs [hunk 4]
```diff
@@ -187,20 +183,55 @@ impl RustProvider {
...
+    fn uses_openssl(app: &App) -> Result<bool> {
+        if let Some(toml_file) = RustProvider::parse_cargo_toml(app)? {
+            if toml_file.dependencies.contains_key("openssl")
+                || toml_file.dev_dependencies.contains_key("openssl")
+            {
+                return Ok(true);
+            }
+        }
+
+        if app.includes_file("Cargo.lock") && app.read_file("Cargo.lock")?.contains("openssl") {
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply src/providers/rust.rs:229]: I think it is clearer to keep them separate. I've added comments to help

File: src/providers/rust.rs [hunk 5]
```diff
@@ -187,20 +183,55 @@ impl RustProvider {
...
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+
+    fn uses_openssl(app: &App) -> Result<bool> {
+        if let Some(toml_file) = RustProvider::parse_cargo_toml(app)? {
+            if toml_file.dependencies.contains_key("openssl")
+                || toml_file.dev_dependencies.contains_key("openssl")
```
Comment (Milo123459) [inline src/providers/rust.rs:220]: Should check build deps as well

**Review by Milo123459 — CHANGES_REQUESTED** [review]

File: examples/rust-openssl/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (coffee-cup) [reply examples/rust-openssl/Cargo.toml:10]: I think having both help for clarity. The main intention of this example is to demonstrate that nixpacks works for rust projects that use either of these dependencies.

**Pull request merged**